File: JAVA/94 - Binary Tree Inorder Traversal.java
/*
 94. Binary Tree Inorder Traversal (M)

 Approach 1 : Iteration
 Time: O(n)
 Space: O(n)

 Approach 2 : Recursion
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> todo = new Stack<TreeNode>();
        while (!todo.empty() || root != null) {
            while (root != null) {
                todo.push(root);
                root = root.left;
            }
            root = todo.pop();
            res.add(root.val);
            root = root.right;
        }
        return res;
    }
}


/* Approach 2 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        DFS(root, res);
        return res;
    }
    
    public void DFS(TreeNode root, List<Integer> res) {
        if (root == null) return;
        DFS(root.left, res);
        res.add(root.val);
        DFS(root.right, res);
    }
}

File: C++/94 - Binary Tree Inorder Traversal.cpp
/*
 94. Binary Tree Inorder Traversal (M)
 
 Approach 1: Recursion
 Time: O(n)
 Space: O(n)

 Approach 2: DFS
 Time: O(n)
 Space: O(n)

*/

// Approach 1
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        inorder(root, res);
        return res;
    }
    
    void inorder(TreeNode* root, vector<int>& res) {
        if (!root) return;
        inorder(root->left, res);
        res.push_back(root->val);
        inorder(root->right, res);
    }
};

// Approach 2
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> todo;
        while (!todo.empty() || root) {
            if (root) {
                todo.push(root);
                root = root->left;
            }
            else {
                root = todo.top();
                todo.pop();
                res.push_back(root->val);
                root = root->right;
            }
        }
        return res;
    }
};